# ImmunoPepper
ImmunoPepper is able to create the set of plausible peptides (or kmers) from a splicing graph, derived from a given
RNA-Seq sample. This peptide set can be personalized with germline and somatic
variants and takes un-annotated introns into account. The comprehensive set of
peptides can then be used further downstream analyses such as domain annotation
or computational immunology.
### Why Use Immunopepper
### Get Started

#### Immunopepper Installation

It is recommended to setup a separate virtual or conda environment.
The basic immunopepper package can be installed from source using:
```
pip install -r requirements.txt -r requirements_dev.txt
make install
```

After that, consult the output of the following command for usage options
```
immunopepper -h
```

#### Basic workflow
We have four working modes for immunopepper `build`, `make_bg`, `diff` and `filter`.

1. `build`: Core part of immunopepper. Traverse splicegraph and get all possible
peptides/kmers.
2. `make_bg`: Integrate multiple kmer files and generate one background kmer file.
3. `diff`: Take as input the foreground kmer file and background kmer file. S
how whether the foreground kmer alsp appears in the background kmer file.
4. `filter`: Apply different filter mechanisms to the given kmer file.

#### Basic options

##### build
- `--samples`: Mandatory argument. the sample names, can specify more than one sample. eg 'TCGA-24-1403 TCGA-24-1002'
- `--output_dir`: Mandatory argument. specify the output directory
- `--ann_path`: Mandatory argument. specify the annotation file, accecpt file format *.gtf*, *.gff* and *.gff3*
- `--splice_path`: Mandatory argument. specify the SplAdder splicegraph path.
- `--ref_path`: Mandatory argument. specify the reference genome.
- `--mutation_mode`: Mandatory argument. specify the mutation mode from {*ref*, *germline*, *somatic* and *somatic_germline*}, default mode *ref*.
- `--kmer`: specify the k for targeted kmer. Default value `0`, which will have no kmer output. The usual setting is `9`.
- `--disable_concat`: If turn on, the concatenated kmer are not considered. The kmers generated by short exons might be missed.
- `--germline`: Specify the germline mutation file path. Mandatory argument if the mutation mode is `germline` or `somatic_and_germline`.
- `--somatic`: Specify the somatic mutation file path. Mandatory argument if the mutation mode is `somatic` or `somatic_and_germline`.
- `--use_mut_pickle`: Check if there is existed pickled mutation dictionary. Use the existed one if yes otherwise process the original mutation files and pickle it.
- `--count_path`: Specify the path of splicegraph count file.
- `--compressed`: Compress the output files with gzip.

Example command line (can replace `ref` to `germline` to consider mutation information)
```
python
immunopepper/main_immuno.py
build
--output_dir
tests/test1/current_output_pos
--ann_path
tests/test1/data/test1pos.gtf
--ref_path
tests/test1/data/test1pos.fa
--splice_path
tests/test1/data/posgraph/spladder/genes_graph_conf3.merge_graphs.pickle
--somatic
tests/test1/data/test1pos.maf
--germline
tests/test1/data/test1pos.vcf
--samples
test1pos
test1neg
--mutation_mode
ref
--kmer
4
--disable_concat
--count_path
tests/test1/data/posgraph/spladder/genes_graph_conf3.merge_graphs.count.hdf5
```

##### make_bg
- `--kmer_files_list`: Mandatory argument. the outputed kmer files generated by build mode, can specify more than one kmer files.
eg 'ref_back_kmer.txt somatic_back_kmer.txt'.
- `--output_file_path`: Mandatory argument. Specify the output integrated background kmer file path.
- `--output_dir`: Mandatoray argument. Specify the directory to store the log file.
- `--verbose`: Specify the level of output. 0 means zero debug information, 2 means the most detailed information.
- `--compressed`: Compress the output files with gzip.

Example command line
```
python
immunopepper/main_immuno.py
make_bg
--kmer_files_list
tests/test1/current_output_pos/test1pos/ref_back_kmer.txt
tests/test1/current_output_pos/test1pos/germline_back_kmer.txt
--output_dir
tests/test1/current_output_pos/
--output_file_path
tests/test1/current_output_pos/test1pos/uniq_back_kmer.txt
--verbose
2
```

##### diff
- `--junction_kmer_file`: Mandatory argument. the foreground junction file path generated by `build` mode. eg `ref_junction_kmer.txt`
- `--bg_file_path`: Mandatory argument. the background kmer file path. Can be the output of `make_bg` mode or external file.
Each line is a new kmer.
- `--output_file_path`: Mandatory argument. Specify the output tsv file path.
- `--output_dir`: Mandatoray argument. Specify the directory to store the log file.
- `--verbose`: Specify the level of output. 0 means zero debug information, 2 means the most detailed information.
- `--compressed`: Compress the output files with gzip.

Example command line
```
python
immunopepper/main_immuno.py
diff
--junction_kmer_file
tests/test1/current_output_pos/test1pos/ref_junction_kmer.txt
--bg_file_path
tests/test1/current_output_pos/test1pos/uniq_back_kmer.txt
--verbose
1
--output_file_path
tests/test1/current_output_pos/test1pos/kmer_result.tsv
--output_dir
tests/test1/current_output_pos
--remove_bg
```

##### filter
- `--junction_kmer_tsv_file`: Mandatory argument. The original kmer tsv files. Generated by `build` mode (without header line)
or by `diff` mode (with header line). It should contain field `cross_junction`, `seg_expr` and `junc_expr`.
- `--cross_junction`: Only output the cross-junction kmers.
- `--seg_expr`: Only output kmers that have segment expression greater than threshold.
- `--seg_expr_thre`: Segment expression threshold. Default 0.
- `--junc_expr`: Only output kmers that have junction expression greater than threshold.
- `--junc_expr_thre`: Junction expression threshold. Default 0.
- `--output_file_path`: Mandatoray argument. Specify the output tsv file path.
- `--output_dir`: Mandatoray argument. Specify the directory to store the log file.
- `--verbose`: Specify the level of output. 0 means zero debug information, 2 means the most detailed information.
- `--compressed`: Compress the output files with gzip.

Example command line
```
python
immunopepper/main_immuno.py
filter
--junction_kmer_tsv_path
tests/test1/current_output_pos/test1pos/kmer_result.tsv
--output_dir
tests/test1/current_output_pos/
--output_file_path
tests/test1/current_output_pos/test1pos/kmer_result_filtered.tsv
--cross_junction
--junc_expr
--verbose
2
```
##### post-processing guidlines
The users can also apply mhc-binding or openMS filter to the output of immunopepper.
###### MHC-Binding
We can use the [NetMHC](http://www.cbs.dtu.dk/services/NetMHC/) interface to do the downstream analysis. `NetMHC` can
predict the peptide-MHC class 1 binding using neural network.

###### Mass spectrum
Mass spectrum data can provide us with another evidence that the kmer exist. [OpenMS](https://www.openms.de) is
a powerful tool for this.
#### Output files
There are 5 files for the `build` mode. `mut_mode` refers to `ref`, `somatic`,  `germline` and `somatic_and_germline`.
- **\[mut_mode\]_back_peptides.fa**: Peptides translated from annotation transcripts. Two lines for one output. The first
line is the transcript ID and the second line is the result peptide.
- **\[mut_mode\]_back_kmer.txt**: kmers generated from **\[mut_mode\]_back_peptides.fa**. There are four columns. First column is
the result kmer, second column is the transcript ID, third column is the average segment expression and final column is the
flag indicating if the kmer is junction kmer. The final column is False for *all* rows in this file.
- **\[mut_mode\]_peptides.fa**: Peptides translated from traversing splicegraph. Two lines for one output. The first
line is the output ID and the second line is the result peptide.
- **\[mut_mode\]_junction_kmer**.txt: kmers generated from **\[mut_mode\]_peptides.fa**. There are one more
column *junction_expr*, which refers to the junction counts for those kmers that spans over
exon junction. For those with *junction_expr* > 0, the flag `is_crossjunction` is True.
- **\[mut_mode\]_metadata.tsv.gz**: Contain details for every junction pairs.

Detail explanation for columns in **\[mut_mode\]_metadata.tsv.gz**
- **output_id**: In the format of \[gene_nama\]:\[first vertex\]_\[second vertex\]:\[somatic variant combination id\]:\[read frame\]. Like `GENE1:0_2:0:1`.
`GENE1` is the gene name, `0_2` means this junction consists of vertex 0 and vertex 2. `0` means there is no
somatic mutation or it is the first case of all somatic mutation combination cases. `2` means the read frame is 2.
- **read_frame**: int (0,1,2). The number of base left to the next junction pair.
- **gene_name**: str. The name of Gene.
- **gene_chr**: str. The Chromosome id where the gene is located.
- **gene_strand**: str ('+', '_'). The strand of gene.
- **mutation_mode**: str ('ref', 'somatic', 'germline', 'somatic_and_germline'). Mutation mode
- **peptide_annotated**: Boolean. Indicate if the junction peptide also appears in the background peptide.
- **junction_peptided**: Boolean. Indicate if the junction also appear in the input annotation file.
- **has_stop_codon**: Boolean. Indicate if there is stop codon in the junction pair.
- **is_in_junction_list**: Boolean. Indicate if the junction pair appear in the given junction whitelist.
- **is_isolated**: Boolean. Indicate if the output peptide is actually translated from a single exon instead of two.
- **variant_comb**: shows the somatic variantion combination used in this line of output. seperate by ';'
    eg. 5;25 means the somatic mutation of position 5 and 25 take effect in this output.
- **variant_seg_expr**: shows the corresponding expression of segments where the corresponding somatic mutation is in.
    eg. 257.0;123.2 means the segment where the somatic mutation in position 5 is in has counts 257.0
- **modified_exons_coor**: Shows exon coordination. Usually we have 4 number start_v1;stop_v1;start_v2;stop_v2. They
    have already absorb reading frame so you can use the coord directly to generate the same output peptide.
- **original_exons_coord**: Shows the original exon coordination.
- **vertex_idx**: shows the vertex id of the given junction. eg 5,6 means this junction pair consists of the fifth and
    sixth vertex.
- **junction_expr**: float. The expression of the junction.
- **segment_expr**: float. The weighted sum of segment expression. We split the junction into segments and compute the segment
    expression with the length-weighted-sum expression.

The `.meta` file is compressed by default in all time. The user can add `--compressed` option
in the input argument to have other files compressed. It is recommended to output in compressed format because
it can save a lot of storage.

The output file for `make_bg` mode is a text file. Each line is a unique kmer.

The output file for `diff` mode is a text file. There is a header line like **\[mut_mode\]_junction_kmer** but
with one more column `is_neo_flag` to indicate if the kmer also exist in the background kmer file. We can also
remove those kmers that exist in the background files with the option `--remove_bg`.

The output file for `filter` mode is a text file also with header line.
#### Real Usecase
We take the real data from the mouse DNA project. Will show how to use the immunopepper to get the new kmers.
In this example, we have two samples `ENCSR000BZG` and `ERR2130621`. We choose `ENCSR000BZG` as the background sample and
`ERR2130621` as the foreground sample. They use the same splicegraph but have different expression
and mutations. We will find what the unique kmers `ERR2130621`  can generate.


- Step 1: Use `build` to get kmers of the two samples in four mutation modes.
```
immunopepper build --mutation_mode ref --samples ENCSR000BZG ERR2130621 --output_dir immunopepper_usecase_out --splice_path immunopepper_usecase.pickle --ann_path immunopepper_usecase.gtf --ref_path genome1.fa --kmer 9 --count_path immunopepper_usecase.count.hdf5
immunopepper build --mutation_mode germline --samples ENCSR000BZG ERR2130621 --output_dir immunopepper_usecase_out --splice_path immunopepper_usecase.pickle --ann_path immunopepper_usecase.gtf --ref_path genome1.fa --kmer 9 --count_path immunopepper_usecase.count.hdf5 --germline immunopepper_usecase.vcf --somatic immunopepper_usecase.maf
immunopepper build --mutation_mode somatic --samples ENCSR000BZG ERR2130621 --output_dir immunopepper_usecase_out --splice_path immunopepper_usecase.pickle --ann_path immunopepper_usecase.gtf --ref_path genome1.fa --kmer 9 --count_path immunopepper_usecase.count.hdf5 --germline immunopepper_usecase.vcf --somatic immunopepper_usecase.maf
immunopepper build --mutation_mode somatic_and_germline --samples ENCSR000BZG ERR2130621 --output_dir immunopepper_usecase_out --splice_path immunopepper_usecase.pickle --ann_path immunopepper_usecase.gtf --ref_path genome1.fa --kmer 9 --count_path immunopepper_usecase.count.hdf5 --germline immunopepper_usecase.vcf --somatic immunopepper_usecase.maf
```

- Step 2: Create background kmer set from the output of sample `ENCSR000BZG`.
Since there is no mutation exist in sample `ENCSR000BZG`, we only consider its output in reference. In addition,
we only consider kmers that have junction expression larger than 0. We can achieve this with `filter` mode and
get file `ref_mode_background_kmer.tsv`. We then use `make_bg` mode to create the background kmer file. Since
the input is just one file. `make_bg` simply takes the first column and output the unique kmers.
```
immunopepper filter --output_dir immunopepper_usecase_out --output_file_path immunopepper_usecase_out/ENCSR000BZG/ref_mode_background_kmer.tsv --junction_kmer_tsv_path immunopepper_usecase_out/ENCSR000BZG/ref_junction_kmer.txt --junc_expr
immunopepper make_bg --kmer_files_list immunopepper_usecase_out/ENCSR000BZG/ref_mode_background_kmer.tsv --output_dir immunopepper_usecase_out --output_file_path immunopepper_usecase_out/background_kmer.txt
```

- Step 3: Remove the background kmers
After getting the background kmers in Step 2, we can remove those kmers from the kmer sets of sample `ERR2130621`. We
can use `diff` to remove those kmers.
```
immunopepper diff --junction_kmer_file  immunopepper_usecase_out/ERR2130621/ref_junction_kmer.txt --bg_file_path immunopepper_usecase_out/background_kmer.txt --output_dir immunopepper_usecase_out --output_file_path immunopepper_usecase_out/ERR2130621/ref_junction_kmer_remove_bg.tsv --remove_bg
immunopepper diff --junction_kmer_file  immunopepper_usecase_out/ERR2130621/germline_junction_kmer.txt --bg_file_path immunopepper_usecase_out/background_kmer.txt --output_dir immunopepper_usecase_out --output_file_path immunopepper_usecase_out/ERR2130621/germline_junction_kmer_remove_bg.tsv --remove_bg
immunopepper diff --junction_kmer_file  immunopepper_usecase_out/ERR2130621/somatic_junction_kmer.txt --bg_file_path immunopepper_usecase_out/background_kmer.txt --output_dir immunopepper_usecase_out --output_file_path immunopepper_usecase_out/ERR2130621/somatic_junction_kmer_remove_bg.tsv --remove_bg
immunopepper diff --junction_kmer_file  immunopepper_usecase_out/ERR2130621/somatic_and_germline_junction_kmer.txt --bg_file_path immunopepper_usecase_out/background_kmer.txt --output_dir immunopepper_usecase_out --output_file_path immunopepper_usecase_out/ERR2130621/somatic_and_germline_junction_kmer_remove_bg.tsv --remove_bg
```

- Step 4: Filter
After removing the background kmers in Step 3, we can add more filters to reduce the numbers of candidate kmers.
For example, we only consider the kmers that have junction expression larger than 0 and also the
segment expression value larger than 2. `filter` mode provides filters based on segment expression and junction expression. We
can set the threshold arbitrarily.
```
immunopepper filter --output_dir immunopepper_usecase_out --output_file_path immunopepper_usecase_out/ERR2130621/ref_junction_kmer_remove_bg_filter.tsv --junction_kmer_tsv_path immunopepper_usecase_out/ERR2130621/ref_junction_kmer_remove_bg.tsv --cross_junction --seg_expr --seg_expr_thre 2
immunopepper filter --output_dir immunopepper_usecase_out --output_file_path immunopepper_usecase_out/ERR2130621/germline_junction_kmer_remove_bg_filter.tsv --junction_kmer_tsv_path immunopepper_usecase_out/ERR2130621/germline_junction_kmer_remove_bg.tsv --cross_junction --seg_expr --seg_expr_thre 2
immunopepper filter --output_dir immunopepper_usecase_out --output_file_path immunopepper_usecase_out/ERR2130621/somatic_junction_kmer_remove_bg_filter.tsv --junction_kmer_tsv_path immunopepper_usecase_out/ERR2130621/somatic_junction_kmer_remove_bg.tsv --cross_junction --seg_expr --seg_expr_thre 2
immunopepper filter --output_dir immunopepper_usecase_out --output_file_path immunopepper_usecase_out/ERR2130621/somatic_and_germline_junction_kmer_remove_bg_filter.tsv --junction_kmer_tsv_path immunopepper_usecase_out/ERR2130621/somatic_and_germline_junction_kmer_remove_bg.tsv --cross_junction --seg_expr --seg_expr_thre 2
```

- Step 5: Aggregate
We get the unique kmers of sample `ERR2130621` in four modes. Now we can aggregate all those kmers.
```
tail -n +2 immunopepper_usecase_out/ERR2130621/*_junction_kmer_remove_bg_filter.tsv | cat | grep -v "==>" | cut -f1 | sort |uniq | grep . > neo_kmer.txt
```
### Tips
- Immunopepper requires the sample name are **exactly** the same in the `splice count file` and `mutation file` and the
given option `--samples` should be those samples. Please make necessary changes to the input files so that immunopepper can work as expected.

- `make_bg`, `diff` and `filter` mode accept the output files of Immunopepper. However, the user
can also add other external input files.
    - `make_bg` assumes the input file has
a header line, seperated with *\t* and kmers are in the first column.
    - `diff` assumes
the foreground kmer file has a header line and that the background kmer file has the
format as the output file of `make_bg`.
    - `filter` assumes the input file has a header
line and with three columns `seg_expr`, `junction_expr` and `is_crossjunction`. It's acceptable if some columns
are missing but the user should not use corresponding filter rules. Otherwise error will happen.
### Contribution Guidelines
### License
### Acknowledgement
