# ImmunoPepper
ImmunoPepper is a software tool for the detection of neoantigens from a splicing graph (possibly derived from RNA-Seq
samples). It generates the set of all theoretically feasible peptide sequences (or kmers) through
direct translation of all walks along the graph. This peptide set can be personalized with germline
and somatic variants and takes all exon-exon junctions present in the splicing graph (even the ones not
part of the reference annotation, but present in the given RNA-Seq sample) into account. The
comprehensive set of peptides can be used subsequently for further downstream analyses, such as
domain annotation or computational immunology.

#TODO: include explanation of background and foreground somewhere
## Get Started

### Installation

It is recommended to set up a separate virtual or conda environment.
The basic ImmunoPepper package can be installed via `pip`:

```
pip install immunopepper
```

Alternatively, ImmunoPepper can also be installed from source using:
```
pip install -r requirements.txt -r requirements_dev.txt
make install
```

After installation, please consult the help screen for further usage options:
```
immunopepper -h
```
### Prerequisites
ImmunoPepper takes a splicing graph as input. This splicing graph has to be generated using the
SplAdder pipeline. Further information about SplAdder is available on its [GitHub
page](https://github.com/ratschlab/spladder) or the [Online
documentation](https://spladder.readthedocs.io/en/latest/).

## Basic workflow
The software has four basic working modes:

1. [`build`](#mode-build): Core part of ImmunoPepper. Traverses the input splice graph and generates all possible
peptides/kmers.
2. [`samplespecif`](#mode-samplespecif): This mode filters from the foreground kmer list the peptides that are also appearing in the background kmer list, and keeps only the novel kmers. It needs to be applied after the build mode.
3. `cancerspecif`:
4. [`mhcbind`](#mode-mhcbind): This mode allows the user to perform MHC binding predictions on the kmers generated by Immunopepper. It is a wrapper tool for different MHC binding affinity prediction tools. It currently contains [NetMHC3](http://www.cbs.dtu.dk/services/NetMHC-3.4/), [NetMHC4](http://www.cbs.dtu.dk/services/NetMHC/), [NetMHCpan](http://www.cbs.dtu.dk/services/NetMHCpan/), [NetMHCIIpan](http://www.cbs.dtu.dk/services/NetMHCIIpan/), [NetMHCcons](http://www.cbs.dtu.dk/services/NetMHCcons/ and MHCflurry. #TODO: add link or command for MHCflurry
### Mode `build`

The following parameters are *mandatory*:
- `--output-dir`: absolute path of the output directory. All output files generated by immunopepper will be saved in this directory.
- `--ann-path`: absolute path for the annotation file. Accepted file formats: *.gtf*, *.gff* and *.gff3*. Annotation files can
be downloaded from various databases such as [GENCODE](https://www.gencodegenes.org/human/)
- `--splice-path`: absolute path of the input SplAdder splice graph.
- `--ref-path`: absolute path of the reference genome file in FASTA format. Reference Please ensure that the reference genome is compatible with the gene annotation file being used.
- `--kmer`: length of the kmers for kmer output.

*Optional submodes* parameters: Commands for conceptual information about the processing.

- `--libsize-extract`: Default = False. Set this parameter to True to generate library sizes and gene quantifications and skip neontigen generation. *Note: If set to True, the program will only output files 3 and 7 of the [output section](#output-files).*
- `--all-read-frames`: Default = False. Set this parameter to True to switch to exhaustive translation and study all possible reading frames instead of just the annotated ones in the annotation file.
- `--count-path`: Default = None. Absolute path for the gene counts file. Used for expression quantification of genes. Format: hdf5.)
- `--output-samples`: Default = None. List of sample names to output. *Note: Names should match the file name of the splice graphs. If not provided all samples are processed and program runs faster.*
- `--heter-code`: Default = 0. It specifies the heterozygous alelle. Options: 0 or 2.

*Optional technical* parameters: Commands for optimization of the software.
- `--compressed`: Default = True. Compress output files in SNAPPY format.
- `--paralell`: Default = 1. Number of threads to be used.
- `--batch-size`: Default = 1000. Number of genes to be processed in each batch.
- `--pickle-samples`: Default = []. List of samples to be pickled. Names should match the sample names of the graph/counts files but an equivalence can be set using --sample-name-map from mutation parameters.

*Optional subset* parameters: Commands to select a subset of the genes to be processed.
- `--process-chr`: Default = None. List of chromosomes to be processed. If not provided all chromosomes are processed.
- `--complexity-cap`: Default = None. Maximum edge complexity of the graph to be processed. If not provided all graphs are processed.
- `--genes-interest`: Default = None. Genes to be processed. Input is a csv file containing a gene id per line, with no header. If not provided all genes are processed.
- `--start-id`: Default = 0. Vertex id of the first gene to be processed. If not provided all genes are processed.
- `--process-num`: Default = 0. Number of genes to be processed. If provided, the first process-num genes are processed.

*Optional output* parameters: Commands to select output formatting and filtering.
- `--skip-annotation`: Default = False. Set this parameter to True to skip the generation of a background file.
- `--keep-tempfiles`: Default = False. Set this parameter to True to keep the temporary files.
- `--libsize-path`: #TODO: not used in build mode
- `--output-fasta`: Default = False. Set this parameter to True to output the foreground peptides in fasta format.
- `--force-ref-peptides`: Default = False. Set this parameter to True to output mutated peptides even if they are the same as the ones in the reference.
- `--kmer-database`: Default = None. Absolute path of a file containing kmers in one column, without header. A file from uniprot or any other standard library can be provided. The kmers provided in this file will not be outputted if found in the foreground peptides. Please note that is a standard proteome is downloaded from an online resource the proteins should be cut into the kmers length selected under `--kmer`.
- `--gtex-junction-path`: Default = None. Absolute path of whitelist junction path. The junctions of this file will be the only ones outputted from the tool. Format: hdf5. #TODO: check the format this file should have
- `--disable-concat`: Default = False. Disable the generation of kmers from combinations of more than 2 exons ((kmers generated from combinations of short exons might be missed). If set to true, kmer generation is faster.
- `--disable-process-libsize`: Default = False. Set to True to generate the libsize file (file 7 from the [output section](#output-files)).

*Optional mutation* parameters: Commands to add somatic and germline mutation information.
- `--mutation-sample`: Default = None. Sample id of the files to which mutations are added. The ids should match the graphs/counts names, but an equivalence can be set with --sample-name-map.
- `--germline`: Default = ''. Absolute path of the germline mutation file. Format: VCF, MAF or h5.
- `--somatic`: Default = ''. Absolute path of the somatic mutation file. Format: VCF, MAF or h5.
- `--sample-name-map`: Default = None. Name mapping to sample names from graphs/counts files. Format: No header. Two columns: *[name of count/graphs file \t name of mutation/pickle file]*. Three columns: *[name of count/graphs file \t name of germline file \t name of somatic file]*.
- `--use-mut-pickle`: Default = False. Set to True to save and use pickled mutation dictionary.

#TODO: Keep this but updated with the new parameters? Leave it for a little bit later and test that it runs properly
Example command line (replace `ref` with `germline` to consider mutation information)
```
immunopepper build \
--output-dir tests/test1/current_output_pos \
--ann-path tests/test1/data/test1pos.gtf \
--ref-path tests/test1/data/test1pos.fa \
--splice-path tests/test1/data/posgraph/spladder/genes_graph_conf3.merge_graphs.pickle \
--somatic tests/test1/data/test1pos.maf \
--germline tests/test1/data/test1pos.vcf \
--samples test1pos \
--mutation-mode ref \
--kmer 4 \
--disable-concat \
--count-path  tests/test1/data/posgraph/spladder/genes_graph_conf3.merge_graphs.count.hdf5 \
--gtex-junction-path tests/test1/data/posgraph/spladder/genes_graph_conf3.test1pos.junction.hdf5
```

### Mode `samplespecif`

The following parameters are *mandatory*:

- `--annot-kmer-files`: List of absolute paths to the annotation kmer files. The files should have the name **\[mut_mode\]_annot_peptides.fa** (Files 1 from the [output section](#output-files))
- `--output-dir`: Path to the output directory.
- `--junction-kmer-files`: List of absolute paths to the sample kmer files. The files are the ones inside the folders **\[mut_mode\]_graph_kmer_JuncExpr** (Files 6 from the [output section](#output-files)) #TODO: also SegmExpr?.
- `--bg-file-path`: Absolute path to the intermediate pooled annotation file. This file is the set of unique kmers in `--annot-kmer-files` files. If the file is not provided it will be generated. Format: One column with header 'kmer'.
- `--output-suffix`: Default = 'no-annot'. Sufflix to be appended to the filtered `--junction-kmer-files`.

The following parameter is *optional*:

- `--remove-bg`: Default = False. Set to True to remove from `--junction-kmer-files`the kmers in `--bg-file-path`. If set to False, a new column `is_neo_flag` will be added to the `--junction-kmer-files` files. The column will contain False if the kmer is in `--bg-file-path` and True otherwise.

#TODO: Keep this but updated with the new parameters or examples?
Example command line:
```
immunopepper make_bg \
--kmer-files tests/test1/current_output_pos/test1pos/ref_back_kmer.txt tests/test1/current_output_pos/test1pos/ref_back_kmer.txt \
--output-dir tests/test1/current_output_pos/ \
--output-file-path tests/test1/current_output_pos/test1pos/uniq_back_kmer.txt \
--verbose 2
```

### Mode `mhcbind`

The following parameters are *mandatory*:

- `mhc-software-path`: Path for the MHC prediction software. Currently supported: netmhc3, netmhc4, netmhcpan, mhcflurry.
- `--argstring`: Complete command line for the MHC prediction tool passed as a string. One should include here the command that will be directly passed to the selected MHC tool. The two *mandatory* arguments are:
  1. `--output-csv`: This argument will contain the path where the MHC prediction tool will save the results.
  2. `--input-peptides-file`: This argument will have the path to the file containing the set of kmers on which MHC binding affinity prediction will be performed. If `--partitioned-tsv`files are provided, an intermediate file will be created and stored under the path `--input-peptides-file`. This intermediate file will contain the set of all unique kmers present in the partitioned files obtained from `cancerspecif` mode. If one does not want to use the output of `cancerspecif` mode for prediction, the path to the file that will be used for prediction will be directly provided under `--input-peptides-list`.

The following parameters are *optional*:
- `--partitioned-tsv`: Default = None. The input to this command is the path to the folder containing the partitioned tsv files from `cancerspecif` mode (output number #TODO:set number of output section(#TODO: add reference)). If this parameter is set the tool will directly accept the files from cancerspecif mode as input.
- `--bind-score-method`: Default = None. Scoring method to filter the MHC tools predictions. E.g. score, affinity, percentile_rank (only for netmhcpan).
- `--bind-score-threshold`: Default = None. Threshold to filter the MHC tools predictions.All the peptides with a score lower than the threshold will be filtered out and only the ones with a score higher than the threshold will be kept.
- `--less-than`: Default = False. If set to True the `--bind-score-threshold` will be considered as an upper bound instead of a lower bound. This means that peptides with a score higher than this threshold will be filtered out.

## Output files

# Mode `build`
There are 10 files for the `build` mode. `mut_mode` refers to `ref`, `somatic`,  `germline` and `somatic_and_germline`.

1. **\[mut_mode\]_annot_peptides.fa**: FASTA file containing background peptides for each gene transcript. The header shows
the transcript ID and the value is the result peptide. Generated only if `--skip-annotation` is set to False. #TODO add example
2. **\[mut_mode\]_annot_kmers**: kmers generated from **\[mut_mode\]_annot_peptides.fa**. There is one column called *kmer* containing each resulting kmer.
Generated only if `--skip-annotation` is set to False. #TODO: Add example
3. **gene_expression_detail**: File containing expression for each gene and each sample. Generated only if `--count-path` file is provided. #TODO: Copy an example of how it looks instead of describing how it looks
4. **\[mut_mode\]_sample_peptides.fa**: FASTA file containing foreground peptides for each gene transcript obtained from traversing splicegraph. The header shows the transcript ID and the value is the result peptide.
Generated only if `--output-fasta` is set to True. #TODO: add example
5. **\[mut_mode\]_graph_kmer_SegmExpr**: Folder containing a file with the expression levels of kmers found in one exon. kmers shown in this file are generated from a single exon and are not located in an exon junction. Format: [kmer, coord, isCrossJunction, junctionAnnotated, readFrameAnnotated, sample names (nº columns = nº samples)]
    - *kmer*: str. The kmer sequence
    - *coord*: str. The genomic coordinates of the kmer. Format: start:end:nan:nan:None:None
    - *isCrossJunction*: Boolean. True if the kmer is located in an exon junction. In this case, all the kmers will get False.
    - *junctionAnnotated*: Boolean. True if the junction kmer is appearing in the annotation file provided under `--ann-path`. In this case, all kmers will get False.
    - *readFrameAnnotated*: Boolean. True if the reading frame was appearing in the annotation file provided under `--ann-path`. The flag value will the False if the reading frame was obtained with the immunopepper tool.
6. **\[mut_mode\]_graph_kmer_JuncExpr**: Folder containing a file with the expression levels of kmers located across exon junctions.
#TODO: check this and add a better example. Format: [kmer, coord, isCrossJunction, junctionAnnotated, readFrameAnnotated]
    - *kmer*: str. The kmer sequence
    - *coord*: str. The genomic coordinates of the kmer. Format: start_e1:end_e1:start_e2:end_e2:start_e3:end_e3. If the kmer is only crossing one junction, start_e3 and end_e3 will be `None`.
    - *isCrossJunction*: Boolean. True if the kmer is located in an exon junction. In this case, all the kmers will get True.
    - *junctionAnnotated*: Boolean. True if the junction kmer is appearing in the annotation file provided under `--ann-path`. If this flag is False, it means that it is a novel kmer.
    - *readFrameAnnotated*: Boolean. True if the reading frame was appearing in the annotation file provided under `--ann-path`. The flag value will the False if the reading frame was obtained with the immunopepper tool.

7. **expression_counts.libsize.tsv**: File containing 75% of expression and total expression for each sample. Format: sample_id, 75% expression, total expression.
Generated only if `--disable-libsize` is set to False and if `--count-path` file is provided.
8. **Annot_IS_SUCCESS**: Empty file indicating that background generation was successful. Generated only if `--skip-annotation` is set to False.
9. **Output_sample_IS_SUCCESS**: Empty file created if the foreground generation was successful.
10. **\[mut_mode\]_sample_peptides_meta**: File containing details for each peptide generated from an exon pair.

Detail explanation for columns in **\[mut_mode\]_sample_peptides_meta**
- **peptide**: str. The peptide sequence for a specific exon pair of the foreground data.
- **id**: In the format of \[gene_name\]:\[first vertex\]_\[second vertex\]:\[variant_id\]:\[translation_start_position\]:\[kmerType\]. Eg: `ENSG00000198515.13:18_15:0:47952701:2-exons`.
`ENSG00000198515.13` is the gene name, `0_2` means this junction consists of vertex 0 and vertex 2. `0` means there is no
somatic mutation or that it is the first case of all somatic mutation combination cases. `47952701 ` is the translation start position. `2-exons` means that the peptide is made by the combination of 2 exons.
- **readFrame**: int (0,1,2). Reading frame used for translating the peptide.
- **readFrameAnnotated**: True if the reading frame was appearing in the annotation file provided under `--ann-path`. The flag value will the False if the reading frame was obtained with the immunopepper tool.
- **geneName**: str. The name of Gene.
- **geneChr**: str. The Chromosome id where the gene is located.
- **geneStrand**: str ('+', '_'). The strand of gene.
- **mutationMode**: str ('ref', 'somatic', 'germline', 'somatic_and_germline'). Mutation mode
- **junctionAnnotated**: #TODO: Remove. Not appearing in metadata
- **hasStopCodon**: Boolean. Indicate if there is stop codon in the junction pair.
- **isJunctionList**: (np.nan, 1, 0). Indicate if the junction pair appear in the given junction whitelist provided under `--gtex-junction-path`.
- **isIsolated**: Boolean. Indicate if the output peptide is actually translated from a single exon instead of two.
- **variantComb**: If mutation files are provided, it shows the somatic mutation combination used in this line of output. Separated by ';' #TODO: is this still separated like this?
    eg. 5;25 means the somatic mutation of position 5 and 25 take effect in this output.
- **variantSegExpr**: If mutation file and count file are provided, this field shows the expression of segments where the somatic mutation is in.
    eg. 257.0;123.2 means the segment where the somatic mutation in position 5 is in has counts 257.0 #TODO: check if this is still like this
- **modifiedExonsCoord**: Coordinates for the exons forming the junction. Usually we have 4 number start_v1;stop_v1;start_v2;stop_v2. It is obtained taking into account the CDS reading frame.
- **originalExonsCoord**: Shows the original exon coordinates obtained from splice graph without taking into account the CDS.
- **vertexIdx**: shows the vertex id of the given junction. Eg: 5,6 means this junction pair consists of the fifth and
    sixth vertex.
- **junctionExpr**: float. The expression of the junction. #TODO: Remove. Not appearing in metadata
- **segment_expr**: float. The weighted sum of segment expression. We split the junction into segments and compute the segment
    expression with the length-weighted-sum expression. #TODO: Remove. Not appearing in metadata
- **kmerType**: str. Shows whether the peptide was translated from 2 or 3 exons.

# Mode `samplespecif`

The output of this mode is a modified version of the files in **\[mut_mode\]_graph_kmer_JuncExpr** and **\[mut_mode\]_graph_kmer_SegmExpr**. #TODO: check if this last folder too.
If `--remove-bg` is set to False, the files will contain a new column called `is_neo_flag`. This flag will be True if the kmer is unique to the foreground data and False if it is also present in the background data.
If `--remove-bg` is set to True, the mode will return the files without the kmers that are common with the background data.

# Mode `mhcbind`

1. **[--input-peptides-file]**: Intermediate file created if `--partitioned-tsv` is provided. It contains the unique set of kmers present in the partitioned tsv files provided, which are the outputs of the cancer specific mode (output number *** #TODO: add number and reference). The file will be stored in the path provided under `--input-peptides-file` in `--argstring`. Format: It will be a csv file, and it will contain a column with the unique kmers, without a header or index.
2. **[--output-csv]**:  Output file generated by the selected MHC tool. Format: csv file, and the exact format will depend on the MHC tool selected. The file will be stored in the path provided under `--output-csv` in `--argstring`.
3. **[--output-csv]_With{`--bind-score-method`}LessLim{`--bind-score-threshold`}.tsv**: File containing the filtered version of **[--output-csv]**. It is generated if filtering options are provided, and `--less-than` is set to True. It will contain kmers that have a score `--bind-score-method` smaller than `--bind-score-threshold`. #TODO: add reference to inner text
4. **[--output-csv]_With{`--bind-score-method`}MoreLim{`--bind-score-threshold`}.tsv**: File containing the filtered version of **[--output-csv]**. It is generated if filtering options are provided, and `--less-than` is set to False. It will contain kmers that have a score `--bind-score-method` bigger than `--bind-score-threshold`. #TODO: add reference to inner text
