# ImmunoPepper
ImmunopPepper is a script which predicts the pepetide sequency from the RNA sequence splicegraph. You can specify the sample and output using the command line below.
### Why Use Immunopepper
### Get Started
#### Immunopepper Installation

It is recommended to setup a separate virtual or conda environment.
The basic immunopepper package can be installed from source using:
```
pip install -r requirements.txt -r requirements_dev.txt
make install
```

After that, consult the output of the following command for usage options
```
immunopepper -h
```

#### Basic workflow
We have four working modes for immunopepper `build`, `make_bg`, `diff` and `filter`.

1. `build`: Core part of immunopepper. Traverse splicegraph and get all possible
peptides/kmers.
2. `make_bg`: Integrate multiple kmer files and generate one background kmer file.
3. `diff`: Add a new column to the kmer tsv file indicating if the kme also appear
in the given background kmer file.
4. `filter`: Apply different filter mechanisms to the given kmer tsv file.

#### Basic options

##### build
- `--samples`: Mandatory argument. the sample names, can specify more than one sample. eg 'TCGA-24-1403 TCGA-24-1002'
- `--output_dir`: Mandatory argument. specify the output directory
- `--ann_path`: Mandatory argument. specify the annotation file, accecpt file format *.gtf*, *.gff* and *.gff3*
- `--splice_path`: Mandatory argument. specify the SplAdder splicegraph path.
- `--ref_path`: Mandatory argument. specify the reference genome.
- `--mutation_mode`: Mandatory argument. specify the mutation mode from {*ref*, *germline*, *somatic* and *somatic_germline*}, default mode *ref*.
- `--kmer`: specify the k for targeted kmer. Default value `0`, which will have no kmer output. The usual setting is `9`.
- `--disable_concat`: If turn on, the concatenated kmer are not considered. The kmers generated by short exons might be missed.
- `--vcf_path`: Specify the vcf file path. Mandatory argument if the mutation mode is `germline` or `somatic_and_germline`.
- `--maf_path`: Specify the maf file path. Mandatory argument if the mutation mode is `somatic` or `somatic_and_germline`.
- `--count_path`: Specify the path of splicegraph count file.

##### make_bg
- `--kmer_files_list`: Mandatory argument. the outputed kmer files generated by build mode, can specify more than one kmer files.
eg 'ref_back_kmer.txt somatic_back_kmer.txt'.
- `--output_file_path`: Mandatory argument. Specify the output integrated background kmer file path.
- `--verbose`: Specify the level of output. 0 means zero debug information, 2 means the most detailed information.

##### diff
- `--junction_kmer_file`: Mandatory argument. the foreground junction file path generated by `build` mode. eg `ref_junction_kmer.txt`
- `--bg_file_path`: Mandatory argument. the background kmer file path. Can be the output of `make_bg` mode or external file.
Each line is a new kmer.
- `--output_file_path`: Specify the output tsv file path.
- `--verbose`: Specify the level of output. 0 means zero debug information, 2 means the most detailed information.
##### filter
- `--kmer_tsv_file`: Mandatory argument. The original kmer tsv files. Generated by `build` mode (without header line)
or by `diff` mode (with header line). It should contain field `cross_junction`, `seg_expr` and `junc_expr`.
- `--cross_junction`: Only output the cross-junction kmers.
- `--seg_expr`: Only output kmers that have segment expression greater than threshold.
- `--seg_expr_thre`: Segment expression threshold. Default 0.
- `--junc_expr`: Only output kmers that have junction expression greater than threshold.
- `--junc_expr_thre`: Junction expression threshold. Default 0.
- `--output_dir`: directory to save filtered kmer file. The output file name will be your specified
filter rules and the original input file name. eg `junc_expr_0_cj_kmer_result.tsv` indicates that the file only
include kmers with cross junction expression greater than 0.
- `--verbose`: Specify the level of output. 0 means zero debug information, 2 means the most detailed information.
### Contribution Guidelines
### License
### Acknowledgement
