# ImmunoPepper
ImmunoPepper is able to create the set of plausible peptides (or kmers) from a splicing graph, derived from a given
RNA-Seq sample. This peptide set can be personalized with germline and somatic
variants and takes un-annotated introns into account. The comprehensive set of
peptides can then be used further downstream analyses such as domain annotation
or computational immunology.
### Why Use Immunopepper
### Get Started
#### Immunopepper Installation

It is recommended to setup a separate virtual or conda environment.
The basic immunopepper package can be installed from source using:
```
pip install -r requirements.txt -r requirements_dev.txt
make install
```

After that, consult the output of the following command for usage options
```
immunopepper -h
```

#### Basic workflow
We have four working modes for immunopepper `build`, `make_bg`, `diff` and `filter`.

1. `build`: Core part of immunopepper. Traverse splicegraph and get all possible
peptides/kmers.
2. `make_bg`: Integrate multiple kmer files and generate one background kmer file.
3. `diff`: Add a new column to the kmer tsv file indicating if the kme also appear
in the given background kmer file.
4. `filter`: Apply different filter mechanisms to the given kmer tsv file.

#### Basic options

##### build
- `--samples`: Mandatory argument. the sample names, can specify more than one sample. eg 'TCGA-24-1403 TCGA-24-1002'
- `--output_dir`: Mandatory argument. specify the output directory
- `--ann_path`: Mandatory argument. specify the annotation file, accecpt file format *.gtf*, *.gff* and *.gff3*
- `--splice_path`: Mandatory argument. specify the SplAdder splicegraph path.
- `--ref_path`: Mandatory argument. specify the reference genome.
- `--mutation_mode`: Mandatory argument. specify the mutation mode from {*ref*, *germline*, *somatic* and *somatic_germline*}, default mode *ref*.
- `--kmer`: specify the k for targeted kmer. Default value `0`, which will have no kmer output. The usual setting is `9`.
- `--disable_concat`: If turn on, the concatenated kmer are not considered. The kmers generated by short exons might be missed.
- `--vcf_path`: Specify the vcf file path. Mandatory argument if the mutation mode is `germline` or `somatic_and_germline`.
- `--maf_path`: Specify the maf file path. Mandatory argument if the mutation mode is `somatic` or `somatic_and_germline`.
- `--count_path`: Specify the path of splicegraph count file.

Example command line (can replace `ref` to `germline` to consider mutation information)
```
python
immunopepper/main_immuno.py
build
--output_dir
tests/test1/current_output_pos
--ann_path
tests/test1/data/test1pos.gtf
--ref_path
tests/test1/data/test1pos.fa
--splice_path
tests/test1/data/posgraph/spladder/genes_graph_conf3.merge_graphs.pickle
--maf_path
tests/test1/data/test1pos.maf
--vcf_path
tests/test1/data/test1pos.vcf
--samples
test1pos
test1neg
--mutation_mode
ref
--kmer
4
--disable_concat
--count_path
tests/test1/data/posgraph/spladder/genes_graph_conf3.merge_graphs.count.hdf5
```

##### make_bg
- `--kmer_files_list`: Mandatory argument. the outputed kmer files generated by build mode, can specify more than one kmer files.
eg 'ref_back_kmer.txt somatic_back_kmer.txt'.
- `--output_file_path`: Mandatory argument. Specify the output integrated background kmer file path.
- `--output_dir`: Mandatoray argument. Specify the directory to store the log file.
- `--verbose`: Specify the level of output. 0 means zero debug information, 2 means the most detailed information.

Example command line
```
python
immunopepper/main_immuno.py
make_bg
--kmer_files_list
tests/test1/current_output_pos/test1pos/ref_back_kmer.txt
tests/test1/current_output_pos/test1pos/germline_back_kmer.txt
--output_dir
tests/test1/current_output_pos/
--output_file_path
tests/test1/current_output_pos/test1pos/uniq_back_kmer.txt
--verbose
2
```

##### diff
- `--junction_kmer_file`: Mandatory argument. the foreground junction file path generated by `build` mode. eg `ref_junction_kmer.txt`
- `--bg_file_path`: Mandatory argument. the background kmer file path. Can be the output of `make_bg` mode or external file.
Each line is a new kmer.
- `--output_file_path`: Mandatory argument. Specify the output tsv file path.
- `--output_dir`: Mandatoray argument. Specify the directory to store the log file.
- `--verbose`: Specify the level of output. 0 means zero debug information, 2 means the most detailed information.
Example command line
```
python
immunopepper/main_immuno.py
diff
--junction_kmer_file
tests/test1/current_output_pos/test1pos/ref_junction_kmer.txt
--bg_file_path
tests/test1/current_output_pos/test1pos/uniq_back_kmer.txt
--verbose
1
--output_file_path
tests/test1/current_output_pos/test1pos/kmer_result.tsv
--output_dir
tests/test1/current_output_pos
--remove_bg
```

##### filter
- `--junction_kmer_tsv_file`: Mandatory argument. The original kmer tsv files. Generated by `build` mode (without header line)
or by `diff` mode (with header line). It should contain field `cross_junction`, `seg_expr` and `junc_expr`.
- `--cross_junction`: Only output the cross-junction kmers.
- `--seg_expr`: Only output kmers that have segment expression greater than threshold.
- `--seg_expr_thre`: Segment expression threshold. Default 0.
- `--junc_expr`: Only output kmers that have junction expression greater than threshold.
- `--junc_expr_thre`: Junction expression threshold. Default 0.
- `--output_file_path`: Mandatoray argument. Specify the output tsv file path.
- `--output_dir`: Mandatoray argument. Specify the directory to store the log file.
- `--verbose`: Specify the level of output. 0 means zero debug information, 2 means the most detailed information.

Example command line
```
python
immunopepper/main_immuno.py
filter
--junction_kmer_tsv_path
tests/test1/current_output_pos/test1pos/kmer_result.tsv
--output_dir
tests/test1/current_output_pos/
--output_file_path
tests/test1/current_output_pos/test1pos/kmer_result_filtered.tsv
--cross_junction
--junc_expr
--verbose
2
```
##### post-processing guidlines
The users can also apply mhc-binding or openMS filter to the output of immunopepper.
###### MHC-Binding
We can use the [NetMHC](http://www.cbs.dtu.dk/services/NetMHC/) interface to do the downstream analysis. `NetMHC` can
predict the peptide-MHC class 1 binding using neural network.

###### Mass spectrum
Mass spectrum data can provide us with another evidence that the kmer exist. [OpenMS](https://www.openms.de) is
a powerful tool for this.
### Contribution Guidelines
### License
### Acknowledgement
